man/en/jubactl.8
.\" Man page generated from reStructuredText.
.
.TH "JUBACTL" "8" " " "" "Jubatus"
.SH NAME
jubactl \- jubactl
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
jubactl [options ...]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
\fBjubactl\fP is a command to manage server processes in distributed environment.
.SS Options
.INDENT 0.0
.IP \(bu 2
\fB[]\fP indicates the default value.
.IP \(bu 2
\fB<zookeeper_list>\fP is a comma\-separated list of \fBhost:port\fP of ZooKeeper servers (e.g., \fB10.0.0.1:2181,10.0.0.2:2181,10.0.0.3:2181\fP).
No spaces between values are allowed.
If you have only one ZooKeeper server, just specify \fBhost:port\fP without comma (e.g., \fB10.0.0.1:2181\fP).
.UNINDENT
.INDENT 0.0
.TP
.B \-c <command>, \-\-cmd <command>
Send specified command to jubavisors registered to ZooKeeper.
\fB<command>\fP should be one of the following.
.TS
center;
|l|l|.
_
T{
Command
T}	T{
Description
T}
_
T{
start
T}	T{
Start Jubatus servers
T}
_
T{
stop
T}	T{
Stop Jubatus servers
T}
_
T{
save
T}	T{
Save the model to directory specified by \fBjubatus_server \-d\fP
T}
_
T{
load
T}	T{
Load the model from directory specified by \fBjubatus_server \-d\fP
T}
_
T{
status
T}	T{
Print the status of servers, proxies and jubavisors
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-s <program>, \-\-server <program>
Executable file of the server program (e.g., \fBjubaclassifier\fP, \fBjubarecommender\fP, ...).
.UNINDENT
.INDENT 0.0
.TP
.B \-n <name>, \-\-name <name>
The instance name, which is a value to uniquely identify a task in the ZooKeeper cluster.
.UNINDENT
.INDENT 0.0
.TP
.B \-t <type>, \-\-type <type>
Type of the server program (e.g., \fBclassifier\fP, \fBrecommender\fP, ...).
.UNINDENT
.INDENT 0.0
.TP
.B \-N <num>, \-\-num <num>
Number of processes in the whole cluster.
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.sp
When \fB0\fP is specified, start 1 process on each jubavisor.
.UNINDENT
.INDENT 0.0
.TP
.B \-z <zookeeper_list>, \-\-zookeeper <zookeeper_list>
List of ZooKeeper server(s).
.sp
If not specified, environment variable \fBZK\fP will be used.
.UNINDENT
.INDENT 0.0
.TP
.B \-i <id>, \-\-id <id>
ID of file name to save or load.
.sp
Effective only when used with \fB\-\-cmd save\fP and \fB\-\-cmd load\fP\&.
.sp
If not specified, the value that specified by \fB\-\-name\fP will be used.
.UNINDENT
.INDENT 0.0
.TP
.B \-B <interface>, \-\-listen_if <interface>
Option given when starting new server process (\fBjubatus_server \-B\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-C <num>, \-\-thread <num>
Option given when starting new server process (\fBjubatus_server \-c\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-T <seconds>, \-\-timeout <seconds>
Option given when starting new server process (\fBjubatus_server \-t\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-D <dirpath>, \-\-datadir <dirpath>
Option given when starting new server process (\fBjubatus_server \-d\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-L <dirpath>, \-\-logdir <dirpath>
Option given when starting new server process (\fBjubatus_server \-l\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-G <log_config>, \-\-log_config <log_config>
Option given when starting new server process (\fBjubatus_server \-g\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-X <mixer>, \-\-mixer <mixer>
Option given when starting new server process (\fBjubatus_server \-x\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-S <seconds>, \-\-interval_sec <seconds>
Option given when starting new server process (\fBjubatus_server \-s\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-I <count>, \-\-interval_count <count>
Option given when starting new server process (\fBjubatus_server \-i\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-Z <seconds>, \-\-zookeeper_timeout <seconds>
Option given when starting new server process (\fBjubatus_server \-Z\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-R <seconds>, \-\-interconnect_timeout <seconds>
Option given when starting new server process (\fBjubatus_server \-I\fP).
.sp
Effective only when used with \fB\-\-cmd start\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-d, \-\-debug
This option is deprecated and is no longer be used.
.UNINDENT
.INDENT 0.0
.TP
.B \-?, \-\-help
Print the brief usage of the command.
.UNINDENT
.SH AUTHOR
PFN & NTT
.SH COPYRIGHT
2011-2017 PFN & NTT
.\" Generated by docutils manpage writer.
.


ChangeLog.rst
Release 1.0.2 - 2017/02/27
--------------------------

* Changes to Supported Systems
    * Ubuntu 16.04 LTS is now officially supported (#1180, #1181)

Release 1.0.1 - 2016/12/26
--------------------------

* Improvements
    * Support OpenCV 2.0 in image extraction plugin (#1168, #1174, #1175)
    * fv_converter plugin loading code moved to Jubatus Core (#1171)
    * Improve Dockerfile (#1170, #1172, #1173)

Release 1.0.0 - 2016/10/31
--------------------------

* New Features
    * Add image extraction plug-in (#843, #1150)
    * Add configuration validation feature (#973, #1144)

* Improvements
    * clustering: change APIs to accept ID-based datum (#1155)
    * Update Dockerfile to use Ubuntu 14.04 (#1146)
    * Improve ``--daemon`` option of server and proxy to detach from console (#1072, #1148)
    * classifier: models are now compatible in standalone and distributed mode (#469, #1153)
    * Change the default log level to INFO instead of DEBUG (#1152)
    * Update example configurations (#1165)
    * Update manpages (#1167)
    * Update packaging tools (#1157)

* Bug fixes
    * graph: fix IDs being reused when loading model file to other server instance in standalone mode (#1149, #1156)

Release 0.9.4 - 2016/08/29
--------------------------

* Improvements
    * Improve error detection of log4cxx config file (#891, #1142)
    * Print DEBUG logs by default (#947, #1143)
    * Support printing logs in UTF-8 (#1078, #1140)
    * Improve comment in wscript (#1138, #1139)

* Bug fixes
    * Fix missing locks in mixer (#1092, #1135)
    * Fix RPM packaging tool to work on RHEL 6.x (#1137)

Release 0.9.3 - 2016/07/25
--------------------------

* Improvements
    * Improve behavior when ZooKeeper log files cannot be reopened (#1026, #1133)
    * Remove Glibc C++ debug flag in debug mode (#910, #1136)
    * Improve packaging tools to remove dependency to Google Code (#980, #1130)

* Bug fixes
    * weight: fix to display config on startup (#1131, #1132)
    * Fix missing lock in get_status of linear_mixer (#1093, #1134)

Release 0.9.2 - 2016/06/27
--------------------------

* Improvements
    * Update README.rst (#1126, #1129)

* Bug fixes
    * Fix `weight_client.hpp` not included in `client.hpp` (#1127, #1128)

Release 0.9.1 - 2016/05/30
--------------------------

* New feature
    * weight: server that returns raw results of fv_converter (#1105, #1116)

* Improvements
    * Add `default.json` config files (#1117, #1120)
    * Add example config file for Anomaly (lof with unlearner) (#1118)
    * Update manpage (#1123)

* Bug fixes
    * Fix thread safetiness issue in save/load RPC (#1113, #1115)
    * Fix `dl` link error on some distributions (#1121, #1122)

Release 0.9.0 - 2016/04/25
--------------------------

* New feature
    * `jubakit <https://github.com/jubatus/jubakit>`_: Python module to easily evaluate precision of models (#1044)

* Changes to Supported Systems
    * Ubuntu 14.04 LTS is now officially supported (#1071)

* Improvements
    * classifier: improve get_labels API to return number of trained instances (#1109)
    * Add example configuration file for Nearest Neighbor (multi-threaded LSH) (#1111)
    * Add example configuration file for Classifier (cosine / euclidean) (#1108)
    * Add example configuration file for Recommender (euclid_lsh with unlearner) (#1108)
    * Configuration in model is used when loading model file on startup (#1081, #1107)
    * Improve compatibility with FreeBSD 9 (#1099)
    * Update manpages (#1112)
    * Code cleanup (#1102)

* Bug fixes
    * Fix jenerator test (#1103)

Release 0.8.9 - 2016/03/28
--------------------------

* Improvements
    * Support extracting specific part-of-speech in MeCab plugin (#1073, #1098)
    * Improved proxy to log timeout errors (#1048, #1095)

* Bug fixes
    * Fix "jubactl -c save" not working (#1094)

Release 0.8.8 - 2016/02/29
--------------------------

* Improvements
    * Support extracting base form in MeCab plugin (#1074, #1085)
    * Compare configuration string semantically on load (#1080, 1087)
    * Add example configuration file for Recommender (inverted_index_euclid with unlearner) (#1090)
    * Fix email address of DEB package (#1082, #1088)

* Bug fixes
    * Fix printing logs multiple times after receiving SIGHUP (#962, #1084, #1086)

Release 0.8.7 - 2016/01/25
--------------------------

* Improvements
    * Add example configuration file for Recommender (MinHash with unlearner) (#1077)
    * Add example configuration file for Recommender and Anomaly (inverted_index_euclid) (#1079)

Release 0.8.6 - 2015/12/25
--------------------------

* Improvements
    * Support morpheme n-gram in mecab plugin (#1054, #1070)

* Bug fixes
    * Fix typos in exception and log messages (#1068, #1069)

Release 0.8.5 - 2015/11/30
--------------------------

* Improvements
    * improve client tests to work on Python 3 (#1043, #1063)
    * improve build script (#1064)
    * recommender: fix typo of calc_similarity function name (#1065)
    * recommender: add example config for LSH with unlearner (#1066)
    * update man pages (#1067)

* Bug fixes
    * Fix RPM package errors (#1059, #1060, #1061, #1062)

Release 0.8.4 - 2015/10/26
--------------------------

* Improvements
    * nearest_neighbor: improve analysis methods performance by eliminating giant lock (#1051)

* Bug fixes
    * Fix loading model files via command line option not working (#1055, #1057)

Release 0.8.3 - 2015/09/28
--------------------------

* Improvements
    * clustering: add seed option to example configuration files (#1047)
    * Fix packaging tools to use new oniguruma URL (#1045, #1046)

* Bug fixes
    * nearest_neighbor: fix argument type (#1042, #1049, #1050)

Release 0.8.2 - 2015/08/31
--------------------------

* Improvements
    * Add Dockerfile (#904)
    * Support add aggregator in jenerator (#1022, #1040)
    * Migrate ux-trie packaging from Google Code to GitHub (#1037)

* Bug fixes

jubatus/server/cmd/jubactl.cpp
// Jubatus: Online machine learning framework for distributed environment
// Copyright (C) 2011 Preferred Networks and Nippon Telegraph and Telephone Corporation.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License version 2.1 as published by the Free Software Foundation.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

#include <locale.h>

#include <string>
#include <vector>
#include <iostream>
#include <cstring>
#include <map>

#include <jubatus/msgpack/rpc/client.h>
#include "jubatus/util/lang/function.h"

#include "jubatus/core/common/exception.hpp"
#include "../third_party/cmdline/cmdline.h"
#include "../common/zk.hpp"
#include "../common/membership.hpp"
#include "../common/logger/logger.hpp"
#include "../framework/server_util.hpp"

static const std::string PROGNAME(JUBATUS_APPNAME "ctl");
static const std::string VERSION(JUBATUS_VERSION);

using std::string;
using std::cout;
using std::endl;
using std::vector;

void usage();
void send2supervisor(
    const string& cmd,
    const string& type,
    const string& name,
    const string& zkhosts,
    const cmdline::parser& argv);
void send2server(
    const string& cmd,
    const string& type,
    const string& name,
    const string& id,
    const string& zkhosts);
void status(const string& type, const string& name, const string& zkhosts);

int main(int argc, char** argv)
try {
  // Explicitly set the current locale to support Unicode output in logs.
  ::setlocale(LC_ALL, "");

  // Configures the logger.
  // We don't provide logging configuration feature for command line tools;
  // just print logs to standard output.
  jubatus::server::common::logger::configure();
  if (!jubatus::server::common::logger::is_configured()) {
    std::cerr << "failed to configure logger" << std::endl;
    ::exit(1);
  }

  cmdline::parser p;

  p.add<std::string>("cmd", 'c',
      "command to send servers (start|stop|save|load|status)", true);
  p.add<std::string>("server", 's',
      "server exec file of learning machine (jubaclassifier, ...)", true);
  p.add<std::string>("name", 'n', "learning machine name", true);
  p.add<std::string>("type", 't',
      "learning machine type (classifier, ...)", true);
  p.add<unsigned int>("num", 'N',
      "num of process in the whole cluster (one on each server when 0)", false);
  p.add<std::string>("zookeeper", 'z',
      "ZooKeeper location; environment variable 'ZK' is used when available",
      false);
  p.add<std::string>("id", 'i',
      "[save|load] id of file name ('name' is used by default)", false);

  // Support framework::server_argv
  p.add<std::string>("listen_if", 'B',
      "[start] bind network interfance", false, "");
  p.add<int>("thread", 'C', "[start] concurrency = thread number", false, 2);
  p.add<int>("timeout", 'T', "[start] time out (sec)", false, 10);
  p.add<std::string>("datadir", 'D',
      "[start] directory to load and save models", false, "/tmp");
  p.add<std::string>("logdir", 'L',
      "[start] directory to output logs (instead of stderr)", false, "");
  p.add<std::string>("log_config", 'G',
      "[start] log4cxx XML configuration file", false, "");
  p.add<std::string>("mixer", 'X',
      "[start] mixer strategy", false, "linear_mixer");
  p.add<int>("interval_sec", 'S', "[start] mix interval by seconds", false, 16);
  p.add<int>("interval_count", 'I',
      "[start] mix interval by update count", false, 512);
  p.add<int>("zookeeper_timeout", 'Z',
      "[start] zookeeper time out (sec)", false, 10);
  p.add<int>("interconnect_timeout", 'R',
      "[start] interconnect time out between servers (sec)", false, 10);

  p.add("debug", 'd', "debug mode (obsolete)");

  p.parse_check(argc, argv);

  string cmd = p.get<std::string>("cmd");
  string name = p.get<std::string>("server") + "/" + p.get<std::string>("name");
  string type = p.get<std::string>("type");

  string zk;
  char* zkenv = std::getenv("ZK");

  if (p.get<std::string>("zookeeper") != "") {
    zk = p.get<std::string>("zookeeper");
  } else if (zkenv) {
    zk = zkenv;
  } else {
    cout << "can't get ZK location: "
        << "set 'ZK' environment or specify '-z <somezkaddrs>" << endl;
    cout << p.get<std::string>("zookeeper") << endl;
    cout << p.usage() << endl;
    exit(1);
  }

  if (cmd == "status") {
    status(type, p.get<string>("name"), zk);
    return 0;
  }

  if (cmd == "start" or cmd == "stop") {
    send2supervisor(cmd, type, name, zk, p);
  }

  if (cmd == "save" or cmd == "load") {
    string id = p.get<std::string>("name");
    if (p.get<std::string>("id") != "") {
      id = p.get<std::string>("id");
    }
    send2server(cmd, type, p.get<string>("name"), id, zk);
  }

  return 0;
} catch (const jubatus::core::common::exception::jubatus_exception& e) {
  std::cout << e.diagnostic_information(true) << std::endl;
}

int do_request(
    const string& cmd,
    const string& name,
    const string& ip_port,
    unsigned int n,
    jubatus::server::framework::server_argv argv) {
  string ip;
  int port;
  jubatus::server::common::revert(ip_port, ip, port);

  msgpack::rpc::client c(ip, port);
  c.set_timeout(10);
  cout << "sending " << cmd << " / " << name << " to "
      << ip_port << "..." << std::flush;

  int r;

  if (cmd == "start") {
    r = c.call(cmd, name, n, argv).get<int>();
  } else {
    r = c.call(cmd, name, n).get<int>();
  }

  if (r != 0) {
    cout << "failed." << endl;
    LOG(ERROR) << "can't do '" << cmd << " " << name << "' in " << ip << " "
        << port << ":" << r;
  } else {
    cout << "ok." << endl;
  }
  return r;
}

void send2supervisor(
    const string& cmd,
    const string& type,
    const string& name,
    const string& zkhosts,
    const cmdline::parser& argv) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));

  vector<string> list;
  jubatus::server::framework::server_argv server_option;

  if (cmd == "start") {
    string path0 = name.substr(name.find_first_of("/") + 1);
    string path;
    jubatus::server::common::build_actor_path(
        path, type, path0.substr(0, path0.find_first_of("/")));
    ls_->create(path);
    path += "/nodes";
    ls_->create(path);

    server_option.bind_if = argv.get<std::string>("listen_if");
    server_option.threadnum = argv.get<int>("thread");
    server_option.timeout = argv.get<int>("timeout");
    server_option.program_name = type;
    server_option.z = zkhosts;
    server_option.name = name;
    server_option.datadir = argv.get<std::string>("datadir");
    server_option.logdir = argv.get<std::string>("logdir");
    server_option.log_config = argv.get<std::string>("log_config");
    server_option.mixer = argv.get<std::string>("mixer");

    server_option.interval_sec = argv.get<int>("interval_sec");
    server_option.interval_count = argv.get<int>("interval_count");
    server_option.zookeeper_timeout = argv.get<int>("zookeeper_timeout");
    server_option.interconnect_timeout = argv.get<int>("interconnect_timeout");
  }

  ls_->list(jubatus::server::common::JUBAVISOR_BASE_PATH, list);

  if (list.empty()) {
    LOG(INFO) << "no server to " << cmd << " " << name;
    exit(-1);
  }

  unsigned int N = argv.get<unsigned int>("num");
  if (N == 0) {
    N = list.size();
  }
  vector<string>::const_iterator it;
  unsigned int n = N / list.size();
  unsigned int m = N % list.size();
  for (it = list.begin(); it != list.end(); ++it) {
    if (m == 0) {
      do_request(cmd, name, *it, n, server_option);
    } else {
      do_request(cmd, name, *it, n + 1, server_option);
      --m;
    }
  }
}

void send2server(
    const string& cmd,
    const string& type,
    const string& name,
    const string& id,
    const string& zkhosts) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));
  std::string path;
  jubatus::server::common::build_actor_path(path, type, name);
  vector<string> list;
  ls_->list(path + "/nodes", list);

  if (list.empty()) {
    LOG(INFO) << "no server to " << cmd << " " << name;
  }

  vector<string>::const_iterator it;
  for (it = list.begin(); it != list.end(); ++it) {
    string ip;
    int port;
    jubatus::server::common::revert(*it, ip, port);

    msgpack::rpc::client c(ip, port);
    c.set_timeout(10);

    cout << "sending " << cmd << " / " << name << " to "
        << *it << "..." << std::flush;

    try {
      if (cmd == "save") {
        c.call(cmd, name, id).get<std::map<string, string> >();
      } else {
        c.call(cmd, name, id).get<bool>();
      }
      cout << "ok." << endl;
    } catch (msgpack::rpc::rpc_error& e) {
      cout << "failed." << endl;
      LOG(ERROR) << "can't do '" << cmd << " " << name << "' in "
          << ip << " " << port;
    }
  }
}

void show(
    jubatus::server::common::lock_service& z,
    const string& path,
    const string& name) {
  vector<string> list;
  z.list(path, list);
  cout << "\033[34mactive " << name << " members:\033[0m" << endl;
  for (size_t i = 0; i < list.size(); ++i) {
    cout << list[i] << endl;
  }
}

void status(const string& type, const string& name, const string& zkhosts) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));
  show(
      *ls_,
      jubatus::server::common::JUBAPROXY_BASE_PATH + "/" + type,
      "jubaproxy");
  show(*ls_, jubatus::server::common::JUBAVISOR_BASE_PATH, "jubavisor");
  std::string path;
  jubatus::server::common::build_actor_path(path, type, name);
  show(*ls_, path + "/nodes", name);
}
